[Mobile] Cart Dropdown

Matt-Yorkley opened the pull request:
#### What? Why?

Closes #4704
Closes #2718 
Closes #4705 
Closes #2706 

<!-- Explain why this change is needed and the solution you propose.
Provide context for others to understand it. -->



#### What should we test?
<!-- List which features should be tested and how. -->

Everything in the cart dropdown epic.

#### Release notes
<!-- Write a line or two to be included in the release notes.
Everything is worth mentioning, because you did it for a reason. -->

Updated UX on cart dropdown for all devices.

<!-- Please assign one category to your PR and delete the others. 
The categories are based on https://keepachangelog.com/en/1.0.0/. -->

luisramos0 commented:
Awesome, I will review soon. Do you plan to fix any of the codeclimate issues? Like the single quotes ones?
Maybe we need to disable some of the rules? Like the ordering of elements?

Matt-Yorkley commented:
I fixed a number of them. I'll have another look and assess some rules.

Matt-Yorkley commented:
I've disabled two rules here:

![Screenshot from 2020-05-30 14-40-13](https://user-images.githubusercontent.com/9029026/83328550-1cc0c680-a284-11ea-81fe-545b5eccdc30.png)

We use double-quoted strings all over the place, and we've decided in the rest of the codebase to prefer double-quoted strings (and disable the same rule in rubocop).

![Screenshot from 2020-05-30 14-44-06](https://user-images.githubusercontent.com/9029026/83328551-1e8a8a00-a284-11ea-9540-480f143c71e8.png)

This rule is only triggered because we extracted breakpoint declarations into a mixin, which I think is an improvement. It would be a total mess if we actually tried to follow this rule everywhere we use breakpoints...

luisramos0 commented:
please don't force-push :pray:

Matt-Yorkley commented on app/assets/stylesheets/darkswarm/cart-page.css.scss:
```diff
@@ -50,7 +53,7 @@
   display: none;
 
-  @media screen and (min-width: 640px) {
+  @include breakpoint(phablet) {
```
Ah you're right, I skimmed over it too quickly. 

Can we wait until after the mobile work is done before doing these sweeping refactors like inverting all the breakpoints?

luisramos0 replied:
Yes yes, as you wish.
I suggest we create the new inverted breakpoints now and start using them as we go.
We need a name for them: media or media-min?

```css
@mixin media($point) {
  @if $point == desktop {
    @media all and (min-width: 1024px) { @content; }
  }
  @else if $point == tablet {
    @media all and (min-width: 768px) { @content; }
  }
  @else if $point == phablet {
    @media all and (min-width: 640px)  { @content; }
  }
  @else if $point == mobile {
    @media all and (min-width: 480px)  { @content; }
  }
}
```

Matt-Yorkley commented on app/views/shared/menu/_cart_sidebar.html.haml:
```diff
@@ -1,8 +1,34 @@
+        %i.ofn-i_009-close
+    .cart-content
+      %table
+        %tr.product-cart{"ng-repeat" => "line_item in Cart.line_items", "id" => "cart-variant-{{ line_item.variant.id }}"}
```
The old layout was using both `Cart.line_items` and `Cart.finalized_line_items` to show items in the current cart and items from previous orders. We're removing that functionality in the new cart to simplify the UX, so we're just using `Cart.line_items` now.

Matt-Yorkley replied:
`Cart.line_items_finalised` is not dead code yet, it's used elsewhere in the app.

luisramos0 replied:
ok, makes sense. now I see, it's used when a second order is placed in the same OC :see_no_evil: 
![image](https://user-images.githubusercontent.com/1640378/83442683-7a017700-a440-11ea-88a4-82a1478c46bc.png)

It's a nice feature but I managed to see a snail in 20secs of trying it by removing the only item in the order that was already completed and then clicking in the link that shows up "empty cart". It doesn't look good. It's unrelated to this PR anyway. I'll ping testers on slack about this.

luisramos0 replied:
anyway, in regards to this PR, all good :+1:

Matt-Yorkley commented on app/assets/stylesheets/darkswarm/shop.css.scss:
```diff
@@ -11,247 +11,170 @@
-      color: #0096ad;
+.open-shop-message {
+  a {
+    color: #0096ad;
```
I just changed the indentation here, but I wasn't changing this code. I'll take a look at the colours now...

Matt-Yorkley replied:
It's also `$teal-500`. :sweat_smile: 

I'll use that instead.

Matt-Yorkley commented on app/serializers/api/variant_serializer.rb:
```diff
@@ -30,4 +30,12 @@ def tag_list
   end
+
+  def thumb_url
+    if object.images.present?
```
I just tried playing with this and it gives the same result either way, whether using `object.product.images.first` or `object.images.first`. :man_shrugging: 

I've updated it to use `object.product.images.first` because in reality we only save images on the product, right? So this is clearer.

luisramos0 replied:
not exactly, we can save an image to a specific variant.
In the new image page there's a field called Version which is the variant.
![image](https://user-images.githubusercontent.com/1640378/83441653-c6e44e00-a43e-11ea-801c-cbb27ccebb3b.png)

But this is not used anywhere and should probably be removed. So Ithink it's ok to use the product image :+1:

sauloperez commented on app/assets/stylesheets/darkswarm/cart-dropdown.css.scss:
```diff
@@ -0,0 +1,115 @@
+    border-spacing: 0;
+    margin: 0;
+
+    tr.product-cart {
```
Is there a reason to specify the `tr`. Avoiding it will have the same result but lower coupling, meaning that things will be a bit more changeable, which is always desirable. The same applies to `tr.bought td` below.

sauloperez commented on app/views/shared/menu/_cart_sidebar.html.haml:
```diff
@@ -0,0 +1,41 @@
+            %img{'ng-src' => '{{ line_item.variant.thumb_url }}'}
+          %td
+            %span {{ line_item.variant.extended_name }}
+            %br
```
why don't we use a block element instead of a `span` so there's no need for a `br`? the need for it clearly shows the `span` doesn't fit the needs.

sauloperez commented on app/assets/stylesheets/darkswarm/cart-page.css.scss:
```diff
@@ -0,0 +1,74 @@
+    margin: 0;
+  }
+
+  .toggle-bought {
```
have you considered using a `button` element instead? we wouldn't need this extra property then. It comes with it.

Matt-Yorkley replied:
Most of these bits are from moving or refactoring existing code, they're not actually changed at all. This rule is for an element in the cart page, which isn't touched in this PR at all...

sauloperez commented on app/assets/stylesheets/darkswarm/shop.css.scss:
```diff
@@ -11,247 +11,172 @@
-        margin-right: 0;
-      }
+  @include breakpoint(tablet) {
+    input.button.right {
```
ufff, this is going to explode soon. A right thing left aligned :exploding_head:

Matt-Yorkley replied:
I've changed the indentation here, but I'm not touching the element this rule applies to. I'd prefer to leave it alone for this PR.

sauloperez commented on app/assets/stylesheets/darkswarm/shop.css.scss:
```diff
@@ -11,247 +11,172 @@
-      button {
-        width: 48%;
+  form {
+    input.small.button.primary.right.add_to_cart {
```
this high specificity selecting the element will bring trouble in the future. Why not simply adding a new class to the specific one we want if we fear unwanted elements would get selected?

sauloperez commented on app/assets/stylesheets/darkswarm/shop.css.scss:
```diff
@@ -11,247 +11,172 @@
-      }
-    }
+    border-bottom: 1px solid #e5e5e5;
+    border-top: 1px solid #e5e5e5;
```
Agree with CodeClimate. A new color deserves a new var, and then, we can question whether it is worth it or it's just 99% similar to an existing one.

Matt-Yorkley commented:
@sauloperez I removed some of the indentation changes. It's probably best not to touch any of that CSS in this PR as I think Maikel will be changing it a lot in other open PRs, and it wasn't strictly necessary or relevant here. All the above comments are now outdated / not relevant.

daniellemoorhead commented:
@Matt-Yorkley I'm partway through testing, you can see the doc here: https://docs.google.com/document/d/1a39-2u5ciC1C30Grp5XAstr5Zmfqa-eicLmqhN5ANaQ/edit?usp=sharing

Its' looking super good! There are a couple of things that I found need fixing that you could sort out in the meantime:

- [ ] The checkout button bumps against a scroll bar on desktop (Mac + Windows) - in the designs there isn't a scroll bar and on iPhone the scroll bar is skinny and transparent so it's not a problem. Interestingly, even when there was no need for scroll the scroll bar was still displaying. It shouldn't, right?

- [ ] The pluralised "items" in the "54 items in your cart" doesn't work when there's just one item. Can you make it "1 item in your cart" and not "1 items in your cart"?

Also, refer to the note I left you about the "Take me shopping!" button on mobile that is described in #2706 which isn't there. It needs to be added, either as part of this PR or as a separate one, whichever works best for you. Just let me know your preference.

I'll finish the remaining testing on Friday when I'm back online. So if you want to sort these out in the meantime go for it 👍

Matt-Yorkley commented:
@daniellemoorhead the `Take me shopping!` button is displayed in some cases and not others. Like if you're already on the a shop page it isn't shown, because we're already shopping...

daniellemoorhead commented:
OK @Matt-Yorkley, next round of testing done 🎉 

Notes [are here](https://docs.google.com/document/d/1a39-2u5ciC1C30Grp5XAstr5Zmfqa-eicLmqhN5ANaQ/edit?usp=sharing)

In #4707:
- [ ] The product name and variant unit name don't truncate as in the design

In #2718:
- [ ] Page/body scroll is not disabled when the cart is open + scrollable
- [ ] Edit cart button doesn't stick to bottom of screen for iPhone / Safari

For both of ☝️  see the document and associated video recordings. Basically, I think the background scroll on iPhone is causing the problem for both of these issues. 

In #4705 
- [ ] Items + close menu doesn't stick to top of screen below cart icon.

Back to you for fixing 🔧

Matt-Yorkley commented:
Okay @daniellemoorhead I've:

- Fixed the cart page button styling regression you noted in Slack
- Added a character limit on product and variant names in cart sidebar
- Made the cart header with close button stick to the top of the cart

I'm not sure about the "page/body scroll when sidebars/modals are open" thing. The issue affects multiple PRs and Mobile UX epics, right? I'll have a crack at it, but it might be best to handle it in a separate PR.

sauloperez commented on app/views/shared/menu/_cart_sidebar.html.haml:
```diff
@@ -16,9 +16,9 @@
             %img{'ng-src' => '{{ line_item.variant.thumb_url }}'}
           %td
-            %span {{ line_item.variant.extended_name }}
+            %span {{ line_item.variant.extended_name | truncate: max_characters }}
```
It'd be better in terms of UX if we used ellipsis too: https://css-tricks.com/snippets/css/truncate-string-with-ellipsis/

Matt-Yorkley replied:
The `truncate` filter is a custom OFN one, and it already inserts an ellipsis.

sauloperez replied:
awesome!

daniellemoorhead commented:
@Matt-Yorkley That was quick!

I’m ok with handling the scroll outside of this PR but it definitely blocks this PR from going luve. iPhone is our most used device and the scroll is a blocker for the cart drop down given how it breaks if there are enough products to cause scrolling. 




> On 12 Jun 2020, at 7:26 pm, Matt-Yorkley <notifications@github.com> wrote:
> 
> ﻿
> Okay @daniellemoorhead I've:
> 
> Fixed the cart page button styling regression you noted in Slack
> Added a character limit on product and variant names in cart sidebar
> Made the cart header with close button stick to the top of the cart
> I'm not sure about the "page/body scroll when sidebars/modals are open" thing. The issue affects multiple PRs and Mobile UX epics, right? I'll have a crack at it, but it might be best to handle it in a separate PR.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

Matt-Yorkley commented:
@daniellemoorhead okay, I'm on it now and I'll include it in this PR. It'll be a re-usable solution that can be used to fix the other cases like modals.

daniellemoorhead commented:
Awesome!

> On 12 Jun 2020, at 7:50 pm, Matt-Yorkley <notifications@github.com> wrote:
> 
> ﻿
> @daniellemoorhead okay, I'm on it now and I'll include it in this PR. It'll be a re-usable solution that can be used to fix the other cases like modals.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

Matt-Yorkley commented:
Okay, that was a slightly tricky problem. It's fixed now and can also be applied when opening modals by including the `BodyScroll` service and calling `BodyScroll.toggle`.

Matt-Yorkley commented:
@luisramos0 @sauloperez one more commit for review: https://github.com/openfoodfoundation/openfoodnetwork/pull/5480/commits/c0d634eba076c9010310062e358344540ac867ff

sigmundpetersen commented:
Build error on cart_spec seems related? The other one is flaky

Matt-Yorkley commented:
Ah, there was a flaky test looking for something in the cart dropdown that doesn't exist anymore. Removed.

daniellemoorhead commented:
Oh noes, still not done 😦 

- [ ] scroll breaks in cart on iPad, and background scroll still happening
- [ ] scrolling on Chrome / android has a weird white section showing up over edit cart / checkout menu.
- [ ] xx items in your cart menu doesn't stick on iPad / safari

Matt-Yorkley commented:
@daniellemoorhead I just checked those issues in Browserstack with iPads in both Safari and Chrome and couldn't replicate them. It seems like it's working fine...?

Was it a specific iPad model?

daniellemoorhead commented:
@Matt-Yorkley I think we can say the iPad problem is a browserstack problem, similar to the android problem that no one could replicate with their own androids. I used my iPad mini 4 and it didn't have the problem, switched to browserstack on the same mini and the problem was there.

So. We consider these a watch and wait problem but not a blocker to getting this PR through.

The only thing remaining is to make sure that the take me shopping button displays on /shops (it wasn't for yuko and I when we looked at it this week).

Matt-Yorkley commented:
If the user is already on /shops and has no selected shop set in their cookies, the button would take them to the /shops page, which is where they already are... So it's not shown.

Matt-Yorkley commented:
Note: rebased on master to pick up changes.

daniellemoorhead commented:
>If the user is already on /shops and has no selected shop set in their cookies, the button would take them to the /shops page, which is where they already are... So it's not shown.

@Matt-Yorkley I go to staging AU. I click on Balsam & Cedar shop, and visit their shop tab. I then return to the shop page. On desktop, it still shows that I'm shopping at that shop. But when I open the cart the take me shopping button doesn't show. This is the use case I'm saying doesn't work, not the one you speak of above.

Matt-Yorkley commented:
@daniellemoorhead all fixed. It needs staging again.

daniellemoorhead commented:
Semaphore failed @Matt-Yorkley - does it need a rebase?

luisramos0 commented:
it's just flakiness on the build, can be staged :+1:

Matt-Yorkley commented:
Rebased to resolve conflicts.

daniellemoorhead commented:
Conflicts don't seem to be resolved, semaphore still failed. Deploying anyway.

daniellemoorhead commented:
Ok, so, the take me shopping button is now displaying when you've visited as shop and then go back to the /shops page. However, it's also showing up on desktop/tablet, which it isn't supposed to do.

@yukoosawa this is something to consider when you do your final design sweep on the cart experience.

@Matt-Yorkley, go ahead and resolve the conflicts and then merge.
